tools/bbc/bbc.cpp

```diff
@@ -199,154 +93,91 @@ std::string CompileFortran(std::string path, Fortran::parser::Options options,
+    errs() << driver.prefix << "semantic errors in " << path << '\n';
+    exitStatus = EXIT_FAILURE;
+    return;
   }
 
   // MLIR+FIR
   fir::NameUniquer nameUniquer;
-  auto burnside = Br::BurnsideBridge::create(
-      semanticsContext.defaultKinds(), &parsing.cooked());
+  auto burnside = Br::BurnsideBridge::create(semanticsContext.defaultKinds(),
+                                             &parsing.cooked());
   fir::KindMapping kindMap{&burnside.getMLIRContext()};
   burnside.lower(parseTree, nameUniquer);
-  mlir::ModuleOp mlirModule{burnside.getModule()};
-  mlir::PassManager pm{mlirModule.getContext()};
-  if (driver.dumpHLFIR) {
-    llvm::errs() << ";== 1 ==\n";
-    mlirModule.dump();
-  }
-  // Run FIR mem2reg and CSE as a pair
+  mlir::ModuleOp mlirModule = burnside.getModule();
+  mlir::PassManager pm = mlirModule.getContext();
+
   pm.addPass(fir::createMemToRegPass());
   pm.addPass(fir::createCSEPass());
-  if (driver.lowerToLLVM) {
-    pm.addPass(fir::createLowerToLoopPass());
-    if (driver.lowerToStd) {
-      pm.addPass(fir::createFIRToStdPass(kindMap));
-    }
-    pm.addPass(mlir::createLowerToCFGPass());
+  pm.addPass(fir::createLowerToLoopPass());
+  pm.addPass(fir::createFIRToStdPass(kindMap));
+  pm.addPass(mlir::createLowerToCFGPass());
+
+  if (EmitLLVM) {
     pm.addPass(fir::createFIRToLLVMPass(nameUniquer));
+    pm.addPass(fir::createLLVMDialectToLLVMPass(OutputFilename + ".ll"));
   }
-  if (driver.lowerToLLVMIR) {
-    pm.addPass(fir::createLLVMDialectToLLVMPass("a.ll"));
-  }
+
   if (mlir::succeeded(pm.run(mlirModule))) {
-    if (driver.dumpFIR) {
-      llvm::errs() << ";== 2 ==\n";
-      mlirModule.dump();
-    }
+    std::error_code ec;
+    raw_fd_ostream out(OutputFilename, ec);
+    if (!ec)
+      mlirModule.print(out);
   } else {
-    llvm::errs() << "oops, pass manager reported failure\n";
+    errs() << "oops, pass manager reported failure\n";
   }
-  return {};
 }
 
-std::string CompileOtherLanguage(std::string path, DriverOptions &driver) {
-  std::string relo{RelocatableName(driver, path)};
-  if (ParentProcess()) {
-    if (!driver.compileOnly && driver.outputPath.empty()) {
-      filesToDelete.push_back(relo);
-    }
-    return relo;
-  }
-  RunOtherCompiler(driver, path.data(), relo.data());
-  return {};
-}
+} // namespace
 
-void Link(std::vector<std::string> &relocatables, DriverOptions &driver) {
-  if (!ParentProcess()) {
-    std::vector<char *> argv;
-    for (size_t j{0}; j < driver.pgf90Args.size(); ++j) {
-      argv.push_back(driver.pgf90Args[j].data());
-    }
-    for (auto &relo : relocatables) {
-      argv.push_back(relo.data());
-    }
-    if (!driver.outputPath.empty()) {
-      char dashO[3] = "-o";
-      argv.push_back(dashO);
-      argv.push_back(driver.outputPath.data());
-    }
-    Exec(argv, driver.verbose);
-  }
-}
-
-int main(int argc, char *const argv[]) {
+int main(int argc, char **argv) {
+  InitLLVM y(argc, argv);
+  (void)y;
```

Comment:
What is the point of this line ?

Reply:
I'll just copy the documentation (since I'm lazy).

// The main() functions in typical LLVM tools start with InitLLVM which does
 // the following one-time initializations:
 //
 //  1. Setting up a signal handler so that pretty stack trace is printed out
 //     if a process crashes.
 //
 //  2. Set up the global new-handler which is called when a memory allocation
 //     attempt fails.
 //
 //  3. If running on Windows, obtain command line arguments using a
 //     multibyte character-aware API and convert arguments into UTF-8
 //     encoding, so that you can assume that command line arguments are
 //     always encoded in UTF-8 on any platform.
 //
 // InitLLVM calls llvm_shutdown() on destruction, which cleans up
 // ManagedStatic objects.

Reply:
The (void)y thing is just a hack around -Werror and unused variables causing spurious build failures.

Reply:
What you will notice now is that when bbc fails, it will magically print a stack traceback.

Reply:
Thanks for explaining this, I had not thought of the unused variable issue.